network/benches/socket_bench.rs

```diff
@@ -306,5 +307,91 @@ fn socket_bench(c: &mut Criterion) {
     c.bench("socket_send_throughput", bench);
 }
 
-criterion_group!(network_benches, socket_bench);
+/// Concurrently dial into server
+fn bench_client_connection<F, T, S>(
+    b: &mut Bencher,
+    concurrency: u64,
+    transport_func: F,
+    server_addr: NetworkAddress,
+) where
+    F: Fn() -> T,
+    T: Transport<Output = S> + Send + Sync + 'static,
+    S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    let mut runtime = Builder::new()
+        .threaded_scheduler()
+        .core_threads(concurrency as usize)
+        .enable_all()
+        .build()
+        .unwrap();
+    b.iter_with_setup(
+        || {
+            let mut futures = vec![];
+            for _ in 0..concurrency {
+                let transport = transport_func();
+                let addr = server_addr.clone();
+                loop {
+                    if let Ok(fut) = transport.dial(addr.clone()) {
+                        futures.push(async move {
+                            match fut.await {
+                                Ok(_socket) => (),
+                                Err(e) => error!("Failed to upgrade {:?}", e),
+                            };
+                        });
+                        break;
+                    }
+                }
+            }
+            futures
+        },
+        |fut| {
+            let mut handles = FuturesUnordered::new();
+            for f in fut {
+                handles.push(runtime.spawn(f));
+            }
+            runtime.block_on(async move { while let Some(Ok(_)) = handles.next().await {} });
+        },
+    );
+}
+
+/// Measure the concurrent connection throughput of the servers, use with socket-bench-server.
+/// TCP_NOISE_ADDR=/ip6/::1/tcp/12345 cargo run --release -p socket-bench-server
+/// TCP_NOISE_ADDR=/ip6/::1/tcp/12345 cargo x bench -p network noise_connections
```

Comment:
Cool, is this the recommended approach for interacting with benchmarks? I would have thought that these would be different bench_functions, like we did in safety_rules.

Reply:
one is to run the server, another is to run the bench. regarding two bench_functions for tcp and noise, I just reused whatever existing code does